fennel/client_tests/branches/test_clone.py

```diff
@@ -0,0 +1,390 @@
+    count: int
+    timestamp: datetime = field(timestamp=True)
+
+    @pipeline(version=1)
+    @inputs(UserInfoDataset)
+    def my_pipeline(cls, user_info: Dataset):
+        return user_info.groupby("country_code").aggregate(
+            Count(window="forever", into_field="count")
+        )
+
+
+@featureset
+class UserInfoFeatureset:
+    user_id: str = feature(id=1)
+    name: str = feature(id=2).extract(field=UserInfoDataset.name, default="None")  # type: ignore
+    age: int = feature(id=3).extract(field=UserInfoDataset.age, default=1)  # type: ignore
+    gender: str = feature(id=4).extract(field=UserInfoDataset.gender, default="None")  # type: ignore
+    country_code: int = feature(id=5).extract(field=UserInfoDataset.country_code, default=1)  # type: ignore
+    email: str = feature(id=6).extract(field=UserInfoDataset.email, default="None")  # type: ignore
+
+
+def _get_changed_datasets(filter_condition):
+    @source(wh.endpoint("UserInfoDataset"))
+    @dataset
+    class UserInfoDataset:
+        user_id: int = field(key=True)
+        name: str
+        age: int
+        gender: str
+        country_code: int
+        email: str
+        timestamp: datetime = field(timestamp=True)
+
+    @dataset
+    class GenderStats:
+        gender: str = field(key=True)
+        count: int
+        timestamp: datetime = field(timestamp=True)
+
+        @pipeline(version=1)
+        @inputs(UserInfoDataset)
+        def my_pipeline(cls, user_info: Dataset):
+            return (
+                user_info.filter(filter_condition)
+                .groupby("gender")
+                .aggregate(Count(window="forever", into_field="count"))
+            )
+
+    return UserInfoDataset, GenderStats
+
+
+@pytest.mark.integration
+@mock
+def test_simple_clone(client):
+    """
+    Cloning should return same datasets and featuresets.
+    """
+    client.sync(
+        datasets=[UserInfoDataset],
+        featuresets=[UserInfoFeatureset],
+    )
+
+    client.clone_branch("test-branch", from_branch="main")
+    assert client.branch == "test-branch"
+
+
+@pytest.mark.integration
+@mock
+def test_clone_after_log(client):
+    """
+    Purpose of this test is to make sure, we copy data also when we clone.
+    """
+    client.sync(
+        datasets=[UserInfoDataset],
+        featuresets=[UserInfoFeatureset],
+    )
+
+    now = datetime.now()
+    data = [
+        {
+            "user_id": 1,
+            "name": "John",
+            "age": 30,
+            "gender": "male",
+            "country_code": 1,
+            "email": "john@fennel",
+            "timestamp": now,
+        },
+    ]
+    df = pd.DataFrame(data)
+    response = client.log("fennel_webhook", "UserInfoDataset", df)
+    assert response.status_code == requests.codes.OK, response.json()
+
+    output = client.get_dataset_df("UserInfoDataset")
+    assert output.shape == (1, 7)
+
+    client.clone_branch("test-branch", from_branch="main")
+    assert client.branch == "test-branch"
```

Comment:
branch should be a getter function rather than accessing the private variables directly

Reply:
added the function